pulsar-broker/src/test/java/org/apache/pulsar/client/api/MutualAuthenticationTest.java

```diff
@@ -0,0 +1,225 @@
+            log.debug("authenticate in client. passed in :{}, send: clientAuthStrings[{}]: {}",
+                dataString, authStep, clientAuthStrings[authStep]);
+            return AuthData.of(clientAuthStrings[authStep].getBytes());
+        }
+    }
+
+    public static class MutualAuthentication implements Authentication {
+        @Override
+        public void close() throws IOException {
+            // noop
+        }
+
+        @Override
+        public String getAuthMethodName() {
+            return "MutualAuthentication";
+        }
+
+        @Override
+        public AuthenticationDataProvider getAuthData(String broker) throws PulsarClientException {
+            try {
+                return new MutualAuthenticationDataProvider();
+            } catch (Exception e) {
+                throw new PulsarClientException(e);
+            }
+        }
+
+        @Override
+        public void configure(Map<String, String> authParams) {
+            // noop
+        }
+
+        @Override
+        public void start() throws PulsarClientException {
+            // noop
+        }
+    }
+
+
+    public static class MutualAuthenticationState implements AuthenticationState {
+        @Override
+        public String getAuthRole() throws AuthenticationException {
+            return "admin";
+        }
+
+        @Override
+        public AuthData authenticate(AuthData authData) throws IOException {
+            String clientData = new String(authData.getBytes(), Charset.forName("UTF-8"));
+            assertEquals(clientData, clientAuthStrings[authStep]);
+            log.debug("authenticate in server. passed in :{}, send: serverAuthStrings[{}]: {}",
+                clientData, authStep, serverAuthStrings[authStep]);
+            AuthData serverData =  AuthData.of(serverAuthStrings[authStep ++].getBytes());
+            return serverData;
+        }
+
+        @Override
+        public AuthenticationDataSource getAuthDataSource() {
+            return null;
+        }
+
+        @Override
+        public boolean isComplete() {
+            return authStep > 1;
+        }
+    }
+
+    public static class MutualAuthenticationProvider implements AuthenticationProvider {
+        @Override
+        public void close() throws IOException {
+        }
+
+        @Override
+        public void initialize(ServiceConfiguration config) throws IOException {
+        }
+
+        @Override
+        public String getAuthMethodName() {
+            return "MutualAuthentication";
+        }
+
+        @Override
+        public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
+            return "admin";
+        }
+
+        @Override
+        public AuthenticationState newAuthState(AuthData authData,
+                                                SocketAddress remoteAddress,
+                                                SSLSession sslSession) {
+            return new MutualAuthenticationState();
```

Comment:
What happens the authData from the first round? I guess that's "init"?

Reply:
right, it is set when newConnectCommand in client side.